gcam-usa to tell

thurber opened the pull request:
`gcam-usa_to_tell.py` uses the query defined in `gcam-usa_to_tell_queries.xml` to extract total electricity demand by year and state from a GCAM-USA output database. The query results are filtered to those with `input == 'electricity domestic supply'`. Columns are renamed in a way to support TELL's expected format. Units are energy in EJ and no conversions are performed in this step.

Please double check that the query and filtering logic is resulting in unambiguous total electricity demand. Please let us know if there are caveats or unspoken assumptions we should be aware of.

Many thanks for your help!

thurber committed:
gcam-usa to tell

crvernon commented:
Hi @thurber, could you state the version of `gcamreader` being used in the README nested in the `gcam_to_tell` dir?  The current version is 1.4 (see https://github.com/JGCRI/gcamreader/tree/main)

thurber committed:
add gcamreader version

thurber commented:
added a note to the README that this should use the latest gcamreader version v1.4

thurber committed:
typo

thurber commented:
of note, this query does not appear to return any data on a gcam-usa v5.3 database

thurber committed:
update filter

thurber commented:
turns out the input shows up in forms like "electricity domestic supply_Jul_day" so i've updated the filter to roll with that

thurber commented:
@pralitp, we aren't confident that this query and filter is appropriate to get the total state level electricity demand -- can you advise how you would go about that if starting from scratch?

cdburley committed:
Update README.MD

pralitp commented on gcam_to_tell/README.MD:
```diff
@@ -25,41 +26,26 @@ TELL needs the following information:
+| GCAM Output | Annual Total Electricity Demand by State | energyFinalConsumBySecEJ | "TBD" | EJ | EJ |
 
+## Notes
+* The GCAM output database has total annual state-level electricity demand by sector. Either GCAM-reader or the 
```
I've verified:  the way you are running the query it is compatible with the IM3 configuration of electricity demands as well as more standard configurations (maybe used in GODEEP?).  But at the very least I would note the query results from the DB may be: state-level electricity demand by sector and **potentially segment (IM3)**.

pralitp commented on gcam_to_tell/README.MD:
```diff
@@ -25,41 +26,26 @@ TELL needs the following information:
+* The GCAM output database has total annual state-level electricity demand by sector. Either GCAM-reader or the 
+associated python script should sum the sector-specific values to create a single total annual state-level electricity 
+demand value for each year.
+* We should double check the units. GCAM seems to simulate demand in EJ, but the output of the GCAM reader is in TWh. We
```
GCAM and therefore GCAM reader would only operate in EJ, never TWh.  Maybe this was an artifact from using `gcam_extractor` which would do the unit conversion?

pralitp commented on gcam_to_tell/README.MD:
```diff
@@ -25,41 +26,26 @@ TELL needs the following information:
+| param     | Parameter name        | str | elecFinalBySecTWh        |
+| x         | Year                  | int | 1975                     |
+| units     | Variable units        | str | Final Electricity by Sector (TWh) |
+| value     | Year                  | flt | 0.510867688888893        |
```
The `Description` on the `value` row needs updating (i.e. shouldn't be `Year`

pralitp commented on gcam_to_tell/gcam-usa_to_tell.py:
```diff
@@ -22,9 +22,11 @@ def get_gcam_electricity_load(
   # aggregate the load
   aggregated = load[
-    (load.input == 'electricity domestic supply') &
+    (load.input.str.startswith('electricity domestic supply'))
```
The query being used is pulling way more data than needed.  Maybe at small scales this isn't a big deal.  But in case it is of concern, you could do the filtering on the query side instead of the python side.  Which of course reduces the volume of data coming out of the XMLDB and into your python script.  (Note: you may want to use a more descriptive title than "inputs by tech" if you do adopt any of the below queries)

The query would then become:
```
<supplyDemandQuery title="inputs by tech">
    <axis1 name="input">input</axis1>
    <axis2 name="Year">demand-physical[@vintage]</axis2>
    <xPath buildList="true" dataName="input" group="false" sumAll="false">
        *[@type='sector']/*[@type='subsector']/*[@type='technology']/*[@type='input' and starts-with(@name, 'electricity domestic supply')]/demand-physical/node()
      </xPath>
    <comments/>
</supplyDemandQuery>
```

And to take it even further, we could skip having to aggregate sectors (and segments) with:
```
<supplyDemandQuery title="inputs by tech">
    <axis1 name="input">input</axis1>
    <axis2 name="Year">demand-physical[@vintage]</axis2>
    <xPath buildList="true" dataName="input" group="false" sumAll="true">
        *[@type='sector' (: collapse :)]/*[@type='subsector' (: collapse :)]/*[@type='technology' (: collapse :)]/*[@type='input' and starts-with(@name, 'electricity domestic supply')]/demand-physical/node()
      </xPath>
    <comments/>
</supplyDemandQuery>
```

pralitp commented on gcam_to_tell/README.MD:
```diff
@@ -25,41 +26,26 @@ TELL needs the following information:
+| subRegion | State acronym         | str | AL                       |
+| param     | Parameter name        | str | elecFinalBySecTWh        |
+| x         | Year                  | int | 1975                     |
+| units     | Variable units        | str | Final Electricity by Sector (TWh) |
```
Maybe I'm missing up but I don't see any unit conversions to `TWh` in this script anymore and further below I see `aggregated['units'] = 'Final Energy by Sector (EJ)'`.

Lets clarify:  what were the output units we want?  Then if we really do want `TWh`, given GCAM will be giving results in EJ, we should apply the conversion in this python script.

pralitp reviewed (changes requested):
Mostly looks good, I've flagged some minor things.  I think the biggest comment to address is clarifying our intentions for Units (and making sure it is working out that way)

pralitp commented:
@thurber I think my review gets at your comments.

> of note, this query does not appear to return any data on a gcam-usa v5.3 database

I think most likely, it was returning results but your `==` filter on the python was too strict.  Is working now that you have starts with?  I've certainly verified running it **directly** works as expected in the IM3 configuration as well as more standard GCAM configurations -- that doesn't mean there couldn't be some _else_ going wrong.

thurber commented:
Thanks @pralitp, your comments make sense. Just to confirm, is this the best way to get at total electricity demand by region? In other words, summing a bunch of values labeled like "electricity domestic supply_Jul_day" feels a little ambiguous and awkward?

@cdburley, I can add a conversion from EJ to TWh into this script if that's the unit you need. In the README you mention needing the demand breakdown by sector, but that's not in here (we sort of did that for GODEEEP to split out transportation). Do you need sector breakdown for IM3?

cdburley committed:
Update README.MD

cdburley commented:
@thurber Correct we do not need it by sector in IM3. Just total. If we can save a step and just output total then that gets my vote. Adding a conversion to TWh in the script seems logical as well.

@pralitp  I've fixed the variable name in the table.

thurber committed:
update query; convert units

thurber commented:
I've updated the query per Pralit's suggestion and added in a conversion from EJ to TWh.

@pralitp, can you reaffirm that this query is the best way to get state level electricity demand?

@cdburley, for EJ to TWh I've multiplied the EJ by 277.7777778 to obtain TWh; can you confirm this is correct?

cdburley commented:
@thurber That conversion from EJ to TWh is correct.

pralitp commented:
> Just to confirm, is this the best way to get at total electricity demand by region? In other words, summing a bunch of values labeled like "electricity domestic supply_Jul_day" feels a little ambiguous and awkward?

Right.  The internals of GCAM are completely generic and ambiguous.  (Almost) Everything is done on naming convention.  From that perspective, yes this is the best way to query the results of interest.  As you suspect, it does mean that queries are liable to "break" with alternative versions of GCAM.  And this is why we always ship a set of queries with GCAM itself so we can manage the link between the code / inputs / configurations and queries and ensure they are consistent.

I don't think you need to address it specifically in this PR, but maybe we can open an issue to come up with a way of at least tracking queries by GCAM version.
